txmempool -/-> validation 1/2: improve performance of check() and remove dependency on validation

glozow opened the pull request:
Remove the txmempool <-> validation circular dependency by removing txmempool's dependency on validation. There are two functions in txmempool that need validation right now: `check()` and `removeForReorg()`. This PR removes the dependencies in `check()`.

This PR also improves the performance of `CTxMemPool::check()` by walking through the entries exactly once, in ascending ancestorcount order, which guarantees that we see parents before children.

glozow committed:
[mempool] remove now-unnecessary code

Remove variables used for keeping track of mempool transactions for
which we haven't processed the parents yet. Since we're iterating in
topological order now, they're always unused.

glozow committed:
MOVEONLY: remove single-use helper func CheckInputsAndUpdateCoins

glozow committed:
[refactor] pass coinsview and height to check()

Removes check's dependency on validation.h

glozow committed:
[mempool] speed up check() by using coins cache and iterating in topo order

No behavior changes.

Before, we're always adding transactions to the "check later" queue if
they have any parents in the mempool. But there's no reason to do this
if all of its inputs are already available from mempoolDuplicate.
Instead, check for inputs, and only mark fDependsWait=true if the
parents haven't been processed yet.

Reduce the amount of "check later" transactions by looking at
ancestors before descendants. Do this by iterating through them in
ascending order by ancestor count. This works because a child will
always have more in-mempool ancestors than its parent.

We should never have any entries in the "check later" queue
after this commit.

glozow committed:
[refactor/bench] make mempool_stress bench reusable and parameterizable

glozow committed:
[bench] Benchmark CTxMemPool::check()

glozow commented:
bench results on master:
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                8.34 |      119,851,301.59 |    0.0% |      0.01 | `MempoolCheck`
```
results on top of this PR:
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                2.13 |      468,844,203.61 |    0.0% |      0.01 | `MempoolCheck`

```

flack commented:
Sorry for being off-topic, but just because I always twitch slightly when I see it: The bench output is already formatted in [Markdown extended syntax](https://www.markdownguide.org/extended-syntax/), so if you just paste it into the comment box, without backticks around it (and without leading whitespace), it renders a bit nicer. And it's easier to do, because less steps :-)

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                8.34 |      119,851,301.59 |    0.0% |      0.01 | `MempoolCheck`

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                2.13 |      468,844,203.61 |    0.0% |      0.01 | `MempoolCheck`

theStack commented on src/bench/mempool_stress.cpp:
```diff
@@ -98,4 +100,27 @@ static void ComplexMemPool(benchmark::Bench& bench)
+        childTxs = static_cast<int>(bench.complexityN());
+    }
+    std::vector<CTransactionRef> ordered_coins = CreateOrderedCoins(det_rand, childTxs, /* min_ancestors */ 5);
+    std::vector<size_t> random_index;
```
This vector doesn't seem to be used anywhere after being filled (aka [WOM](https://en.m.wikipedia.org/wiki/Write-only_memory_(joke)))?

glozow replied:
oops :sweat_smile: was originally going to use this to evict transactions in random order, will remove

theStack reviewed (commented):
Concept ACK

jnewbery commented on src/txmempool.cpp:
```diff
@@ -751,6 +745,10 @@ void CTxMemPool::check(CChainState& active_chainstate) const
         assert(it->GetSizeWithAncestors() == nSizeCheck);
         assert(it->GetSigOpCostWithAncestors() == nSigOpCheck);
         assert(it->GetModFeesWithAncestors() == nFeesCheck);
+        // Sanity check: we are walking in ascending ancestor count order. The non-contextual
+        // check() will have ensured that this count is accurate.
```
I don't understand what "non-contextual check()" is referring to here.

glozow replied:
Leftover from previous iteration of the PR - removed now

jnewbery commented on src/txmempool.cpp:
```diff
@@ -693,28 +686,29 @@ void CTxMemPool::check(CChainState& active_chainstate) const
                 const CTransaction& tx2 = it2->GetTx();
                 assert(tx2.vout.size() > txin.prevout.n && !tx2.vout[txin.prevout.n].IsNull());
-                fDependsWait = true;
+                // Sanity check: all parents should have been checked before their children, which
```
This comment could explain why parents have been checked before their children (because we sorted by depth before iterating through entries).

glozow replied:
Done

jnewbery commented on src/txmempool.cpp:
```diff
@@ -671,14 +672,6 @@ void CTxMemPool::clear()
-    assert(fCheckResult);
-    UpdateCoins(tx, mempoolDuplicate, std::numeric_limits<int>::max());
-}
 
```
Remove extra newline :)

glozow replied:
Fixed

jnewbery commented on src/txmempool.cpp:
```diff
@@ -693,28 +686,29 @@ void CTxMemPool::check(CChainState& active_chainstate) const
-                fDependsWait = true;
+                // Sanity check: all parents should have been checked before their children, which
+                // means its outputs must be available in the coins cache.
+                assert(mempoolDuplicate.HaveCoin(txin.prevout));
```
Can this assert and the `assert(active_coins_tip.HaveCoin(txin.prevout));` in the else branch below now be merged into a single `assert(mempoolDuplicate.HaveCoin(txin.prevout));`?

If I understand correctly, the `mempoolDuplicate` coins cache is a copy of the `active_chains_tip` cache, so should have all of the coins in the UTXO set. That means `mempoolDuplicate` should have all of the coins that are being spent by each transaction as we iterate through the entries in depth order.

```diff
--- a/src/txmempool.cpp
+++ b/src/txmempool.cpp
@@ -706,13 +706,11 @@ void CTxMemPool::check(CChainState& active_chainstate) const
             if (it2 != mapTx.end()) {
                 const CTransaction& tx2 = it2->GetTx();
                 assert(tx2.vout.size() > txin.prevout.n && !tx2.vout[txin.prevout.n].IsNull());
-                // Sanity check: all parents should have been checked before their children, which
-                // means its outputs must be available in the coins cache.
-                assert(mempoolDuplicate.HaveCoin(txin.prevout));
                 setParentCheck.insert(*it2);
-            } else {
-                assert(active_coins_tip.HaveCoin(txin.prevout));
             }
+            // Sanity check: all parents should have been checked before their children, which
+            // means its outputs must be available in the coins cache.
+            assert(mempoolDuplicate.HaveCoin(txin.prevout));
             // Check whether its inputs are marked in mapNextTx.
             auto it3 = mapNextTx.find(txin.prevout);
             assert(it3 != mapNextTx.end());
```

glozow replied:
Yes, it can, but I thought people might find it scary. I've grouped this and the !IsCoinbase check you suggested into one "simplify logic" commit.

jnewbery commented on src/txmempool.cpp:
```diff
@@ -769,24 +767,12 @@ void CTxMemPool::check(CChainState& active_chainstate) const
-        }
+        TxValidationState dummy_state; // Not used. CheckTxInputs() should always pass
+        CAmount txfee = 0;
+        bool fCheckResult = tx.IsCoinBase() || Consensus::CheckTxInputs(tx, dummy_state, mempoolDuplicate, spendheight, txfee);
```
I know you just moved this code, but I think we should remove `tx.IsCoinBase()` from here, and add an `assert(!tx.IsCoinBase());`. The mempool shouldn't contain coinbase transactions.

No need to do it as part of this PR. Can be a follow-up.

glozow replied:
Agree, added.

jnewbery reviewed (commented):
utACK 4bc0c8066e79f0d95053bcc2449a7d354eaa9c25

I've left a few nitty style comments inline.

I found the ordering of the commits:

- _[mempool] speed up check by using coins cache and iterating in topo order_
- _[mempool] assert that check() always goes in topological order_
- _[mempool] remove code needed for unsorted mempool checking_

to be a little bit confusing. In the intermediate commits, you have things like:

```
                assert(mempoolDuplicate.HaveCoin(txin.prevout));
                if (!mempoolDuplicate.HaveCoin(txin.prevout)) fDependsWait = true;

```

which you then remove in the subsequent commit. It's fine, but it's a bit more mental overhead for reviewers.

I think a more intuitive progression of commits would be to squash those three commits into two:

- the first commit just changes the for loop to use `GetSortedDepthAndScore()` and adds the `prev_ancestor_count` sanity check.
- the second commit removes the `waitingOnDependants` code since the iteration is now in depth order.

Is the next step here to change `check()` to take a `const CCoinsViewCache&` and `int64_t height` and remove the dependency on validation entirely?

glozow committed:
[mempool] simplify some check() logic

No transaction in the mempool should ever be a coinbase.

Since mempoolDuplicate's backend is the chainstate coins view, it should
always contain the coins available.

glozow committed:
[validation/mempool] use Spend/AddCoin instead of UpdateCoins

UpdateCoins is an unnecessary dependency on validation. All we need to
do is add and remove coins to check inputs. We don't need the extra
logic for checking coinbases and handling TxUndos.

Also remove the wrapper function in validation.h which constructs a
throwaway TxUndo object before calling UpdateCoins because it is now
unused.

glozow commented on src/bench/mempool_stress.cpp:
```diff
@@ -91,4 +100,27 @@ static void ComplexMemPool(benchmark::Bench& bench)
+    int childTxs = 2000;
+    if (bench.complexityN() > 1) {
+        childTxs = static_cast<int>(bench.complexityN());
+    }
```
did ternary operator on one line

glozow commented on src/bench/mempool_stress.cpp:
```diff
@@ -91,4 +100,27 @@ static void ComplexMemPool(benchmark::Bench& bench)
+    if (bench.complexityN() > 1) {
+        childTxs = static_cast<int>(bench.complexityN());
+    }
+    std::vector<CTransactionRef> ordered_coins = CreateOrderedCoins(det_rand, childTxs, /* min_ancestors */ 5);
```
Done

glozow commented on src/bench/mempool_stress.cpp:
```diff
@@ -91,4 +100,27 @@ static void ComplexMemPool(benchmark::Bench& bench)
+    random_index.resize(ordered_coins.size());
+    std::iota(random_index.begin(), random_index.end(), 0);
+    Shuffle(random_index.begin(), random_index.end(), FastRandomContext());
+    auto extra_args = {"-checkmempool=1"}; // Always check
```
Done

glozow commented on src/bench/mempool_stress.cpp:
```diff
@@ -91,4 +100,27 @@ static void ComplexMemPool(benchmark::Bench& bench)
+    CTxMemPool pool;
+    LOCK2(cs_main, pool.cs);
+    for (auto& tx : ordered_coins) AddTx(tx, pool);
+    bench.run([&]() NO_THREAD_SAFETY_ANALYSIS {
```
Agree, done

glozow commented on src/bench/mempool_stress.cpp:
```diff
@@ -6,7 +6,9 @@
 #include <txmempool.h>
+#include <validation.h>
 
+#include <numeric>
```
Done, sorry for the sloppiness

glozow commented:
@flack good tip, fixed

@theStack @jnewbery thanks for your reviews, I think I've incorporated all suggestions :)

> Is the next step here to change check() to take a const CCoinsViewCache& and int64_t height and remove the dependency on validation entirely?

Yes I thought I had included that commit here but apparently I hadn't - last push added it.

jnewbery commented on src/txmempool.cpp:
```diff
@@ -671,16 +672,7 @@ void CTxMemPool::clear()
-    CAmount txfee = 0;
-    bool fCheckResult = tx.IsCoinBase() || Consensus::CheckTxInputs(tx, dummy_state, mempoolDuplicate, spendheight, txfee);
-    assert(fCheckResult);
-    UpdateCoins(tx, mempoolDuplicate, std::numeric_limits<int>::max());
```
After this commit, `UpdateCoins(const CTransaction& tx, CCoinsViewCache& inputs, int nHeight)` is no longer used:

```diff
diff --git a/src/validation.cpp b/src/validation.cpp
index 4504d2ca0a..c552577229 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -1225,6 +1225,7 @@ void CChainState::InvalidBlockFound(CBlockIndex* pindex, const BlockValidationSt
     }
 }
 
+/** Apply the effects of this transaction on the UTXO set represented by view */
 void UpdateCoins(const CTransaction& tx, CCoinsViewCache& inputs, CTxUndo &txundo, int nHeight)
 {
     // mark inputs spent
@@ -1240,12 +1241,6 @@ void UpdateCoins(const CTransaction& tx, CCoinsViewCache& inputs, CTxUndo &txund
     AddCoins(inputs, tx, nHeight);
 }
 
-void UpdateCoins(const CTransaction& tx, CCoinsViewCache& inputs, int nHeight)
-{
-    CTxUndo txundo;
-    UpdateCoins(tx, inputs, txundo, nHeight);
-}
-
 bool CScriptCheck::operator()() {
     const CScript &scriptSig = ptxTo->vin[nIn].scriptSig;
     const CScriptWitness *witness = &ptxTo->vin[nIn].scriptWitness;
diff --git a/src/validation.h b/src/validation.h
index b2282828ce..caa0832dd3 100644
--- a/src/validation.h
+++ b/src/validation.h
@@ -229,9 +229,6 @@ PackageMempoolAcceptResult ProcessNewPackage(CChainState& active_chainstate, CTx
                                                    const Package& txns, bool test_accept)
                                                    EXCLUSIVE_LOCKS_REQUIRED(cs_main);
 
-/** Apply the effects of this transaction on the UTXO set represented by view */
-void UpdateCoins(const CTransaction& tx, CCoinsViewCache& inputs, int nHeight);
-
 /** Transaction validation functions */
 
 /**
```

jnewbery commented:
> > Is the next step here to change check() to take a const CCoinsViewCache& and int64_t height and remove the dependency on validation entirely?

> Yes I thought I had included that commit here but apparently I hadn't - last push added it.

It might actually be better to defer that commit until after #23173 so fewer changes are needed to net_processing. Entirely up to you.

jnewbery commented:
reACK 082c5bf099c64e3d27abe9b68a71ce500b693e7e

SachinMeier commented on src/bench/mempool_stress.cpp:
```diff
@@ -79,6 +76,17 @@ static void ComplexMemPool(benchmark::Bench& bench)
+static void ComplexMemPool(benchmark::Bench& bench)
+{
+    FastRandomContext det_rand{true};
+    int childTxs = 800;
+    if (bench.complexityN() > 1) {
+        childTxs = static_cast<int>(bench.complexityN());
+    }
```
This could be simplified to: 
`const int childTxs = bench.complexityN() > 1 ? static_cast<int>(bench.complexityN()) : 800;`
as it is on line 105

rajarshimaitra reviewed (commented):
tACK https://github.com/bitcoin/bitcoin/pull/23157/commits/082c5bf099c64e3d27abe9b68a71ce500b693e7e

Bench comparison

#### Optimized
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                2.39 |      417,806,513.20 |    0.6% |      0.01 | `MempoolCheck`
|                2.39 |      419,189,093.49 |    0.6% |      0.01 | `MempoolCheck`
|                2.39 |      419,188,854.66 |    0.6% |      0.01 | `MempoolCheck`
|                2.39 |      419,049,672.97 |    0.3% |      0.01 | `MempoolCheck`

#### Non-Optimized
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               12.61 |       79,317,536.11 |    6.8% |      0.01 | `MempoolCheck`
|               12.20 |       81,995,745.32 |    9.2% |      0.01 | `MempoolCheck`
|               12.22 |       81,862,095.21 |    1.9% |      0.01 | `MempoolCheck`
|               12.17 |       82,167,397.00 |    0.8% |      0.01 | `MempoolCheck`


It seems I am observing only 6x speedup. But that's probably my local benching issue.

theStack reviewed (commented):
Not sure what's different or wrong on my system (running OpenBSD 6.9/amd64), but the benchmark doesn't show any noticable difference after the optimization in `CTxMemPool::check`, it even seems to be worse.

at commit https://github.com/bitcoin/bitcoin/pull/23157/commits/30e240f65e69c6dffcd033afc63895345bd51f53 (with original `check` as on master branch):
```
$ ./src/bench/bench_bitcoin -filter=MempoolCheck
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              100.66 |        9,934,356.36 |    1.6% |      0.01 | `MempoolCheck`

at commit https://github.com/bitcoin/bitcoin/pull/23157/commits/54c6f3c1da01090aee9691a2c2bee0984a054ce8 (with optimized `check`):
```
$ ./src/bench/bench_bitcoin -filter=MempoolCheck
```

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              107.36 |        9,314,108.15 |    4.4% |      0.01 | `MempoolCheck`

Will try on a Debian Linux machine later to compare.

// EDIT: It seems like the difference in benchmark results is almost only caused by the very last commit.

glozow commented:
>  It seems like the difference in benchmark results is almost only caused by the very last commit.

That sounds strange to me, but I guess it's plausible that the bench only sees a speedup because of the parameter change rather than ordered traversal. Even so, I'd want to remove all that unnecessary code...

theStack commented:
> That sounds strange to me, but I guess it's plausible that the bench only sees a speedup because of the parameter change rather than ordered traversal. Even so, I'd want to remove all that unnecessary code...

Agree, removing unnecessary code is always a good thing! I didn't intend to sound concept NACKy, this was just a hint that the posted benchmark results have to be taken with a grain of salt :) will likely code-review this PR within the next days.

theStack reviewed (approved):
Code-review ACK 082c5bf099c64e3d27abe9b68a71ce500b693e7e

mjdietzx reviewed (commented):
Post merge ACK 082c5bf099c64e3d27abe9b68a71ce500b693e7e

Warming up to review some of your open PRs. Just one question / possible nit. You are KILLING it with these PRs, well done 💪
